### PR TITLE
Type 'apply' method of plugins

### PR DESCRIPTION
**Pull request by ooflorent**
**What kind of change does this PR introduce?**

types

**Did you add tests for your changes?**

no

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

- Removed option `output.devtoolLineToLine`
- Removed option `lineToLine` for `SourceMapDevToolPlugin`
- Replaced `compilation.hooks.normalModuleLoader` by `NormalModule.getCompilationHooks(compilation).loader`

### CODE REVIEW

File: lib/Module.js [hunk 1]
```diff
@@ -101,8 +101,13 @@ class Module extends DependenciesBlock {
...
 		this.buildInfo = undefined;
 
-		/** @type {boolean} */
+		// TODO refactor this -> options object filled from Factory
 		this.useSourceMap = false;
+		this.lineToLine = false;
+
+		// TODO figure out if this should be defined here instead of `NormalModule`.
+		// Without this, type checking fails because the hooks use `Module`.
+		this.resource = undefined;
```
Comment (ooflorent) [inline lib/Module.js]: I think we should decide what to do here.
Most hooks use `Module` but a lot of plugins assume `NormalModule`.

File: lib/NormalModule.js [hunk 2]
```diff
@@ -115,11 +115,6 @@ class NormalModule extends Module {
 		/** @private @type {Map<string, CachedSourceEntry>} */
 		this._cachedSources = new Map();
 
-		// Options for the NormalModule set by plugins
-		// TODO refactor this -> options object filled from Factory
-		this.useSourceMap = false;
-		this.lineToLine = false;
```
Comment (ooflorent) [inline lib/NormalModule.js]: Moved to `Module.js`

File: lib/Module.js (hunk 1 above)
Comment (sokra) [reply lib/Module.js]: These cases are actually bugs. Not every Module has `resource`. When used it should check for `instanceof NormalModule`.

Comment (sokra) [reply lib/Module.js]: Same for `lineToLine`, but I guess `lineToLine` can be removed in webpack 5 anyway. Nobody uses this...

File: lib/LoaderOptionsPlugin.js [hunk 3]
```diff
@@ -38,7 +39,11 @@ class LoaderOptionsPlugin {
 		compiler.hooks.compilation.tap("LoaderOptionsPlugin", compilation => {
 			compilation.hooks.normalModuleLoader.tap(
 				"LoaderOptionsPlugin",
-				(context, module) => {
+				(context, m) => {
+					if (!(m instanceof NormalModule)) {
```
[the remark this answers is not in the document]
Comment (ooflorent) [reply lib/LoaderOptionsPlugin.js]: Right. I'll do this.

File: lib/Compilation.js [hunk 4]
```diff
@@ -140,6 +141,11 @@ const { arrayToSetDeprecation } = require("./util/deprecation");
  * @property {boolean=} noChunkHash
  */
 
+// TODO webpack 5: remove
```
Comment (ooflorent) [inline lib/Compilation.js]: Should be webpack 6

**Comment by sokra** [comment]
Thanks

**Pull request merged**